Release/v2.21.0

thallgren opened the pull request.

thallgren committed:
Remove deprecated test-vpn command.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update the telepresence uninstall command.

- Remove the since long deprecated `--everything`.
- Deprecate the `--agents` option. It's now the default.
- Add command completion.
- Fix documentation around traffic-agent sidecar (and uninstall).

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Maximize timeouts.connectivityCheck to 5s.

Using a higher value is pointless and just leads to root daemon
timeouts.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Some minor documentation fixes.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.0-rc.0

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix failing docs build.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Use relative doc links in CHANGELOG.yml

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add more links to the release notes.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let root connect timeout be half of timeouts.trafficManagerConnect.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Remove the gather-traces and upload-traces commands.

The ability to collect trace has been removed. The code was complex and
has not been well maintained since its inception in 2022. We have seen
no indication that it has ever been used.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Move the telepresence-intercept-env data into traffic-manager configmap

There's really no need for two configmaps that store configuration data
for the traffic manager. The traffic-manager configmap is also watched,
which saves a lot of k8s API calls.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add section on helm install from ghcr.io.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Improve changelog wording on VPN conflict avoidance.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Improve troubleshooting advice on how to avoid init-containers

Closes #3730

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Let routes belonging to an allowed conflict be added as a static route

The `allowConflicting` setting didn't always work on Linux because the
conflicting subnet was just added as a link to the TUN device, and
therefore didn't get subjected to routing rule used to assign priority
to the given subnet.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Replace references to jmalloc/echo server to avoid Docker pull limits

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on docs/howtos/intercepts.md:
````diff
@@ -65,7 +65,7 @@ network telepresence, and remote mounts must be made relative to a specific moun
 4. Intercept all traffic going to the application in your cluster:
     ```
-    telepresence intercept <workload-name> --port [<local-port>][:<remote-port>] --env-file <path-to-env-file>`.
+    telepresence intercept &lt;workload-name&gt; --port [&lt;local-port&gt;][:&lt;remote-port&gt;] --env-file &lt;path-to-env-file&gt;`.
````
any reason for these HTML character entities?
They aren't really compliant with GFM, so the preview on GH is somewhat broken - less than/greater than inside of a fenced code block should already be escaped as `&lt;` and `&gt;`.
Do we need that due to the telepresence.io app?

thallgren replied:
Yeah, I made that change after a build failure from docusaurus suggesting that this was treated as mdx, which it shouldn't be given the extension. I'll investigate this further, because the preview on GH should be as good as it can be.

P0lip commented on cmd/traffic/cmd/manager/config/config.go:
```diff
@@ -115,9 +124,27 @@ func (c *config) refreshFile(ctx context.Context, data map[string]string) {
+
+	c.agentEnv = AgentEnv{}
+	if yml, ok := data[agentEnvConfigFileName]; ok {
+		data, err := yaml.YAMLToJSON([]byte(yml))
+		if err == nil {
+			err = json.Unmarshal(data, &c.agentEnv)
+		}
```
```
nit - don't need to address it now, but at some point. YAML 1.2 much compliant with JSON assuming JSON or core schemas are used (core is the recommended default for processors). There are a few differences that should be irrelevant in this context.
As such, one can do
```suggestion
		data, err := yaml.Unmarshal([]byte(yml))
```
to handle both file types.

thallgren replied:
There's no yaml that can handle the JSON functionality in the json package that we're using now. (such as the `json` tags and `omitzero` See https://github.com/go-json-experiment/json for more info.

P0lip commented on pkg/client/cli/cmd/uninstall.go:
```diff
@@ -2,88 +2,76 @@ package cmd
-		Short:             "Uninstall telepresence agents",
-		RunE:              ui.run,
-		ValidArgsFunction: cobra.NoFileCompletions,
+		Use:   "uninstall [flags] <workloads...>",
```
it might be good to mention that positional argument in a changelog perhaps?

thallgren replied:
Yep, will address that.

P0lip commented on pkg/client/userd/trafficmgr/session.go:
```diff
@@ -1037,9 +1034,10 @@ func (s *session) connectRootDaemon(ctx context.Context, nc *rootdRpc.NetworkCon
 		for attempt := 1; ; attempt++ {
 			var rootStatus *rootdRpc.DaemonStatus
-			tCtx, tCancel := context.WithTimeout(ctx, 15*time.Second)
+			tCtx, tCancel := context.WithTimeout(ctx, tmTimeout/2)
```
I assume we divide by 2 to effectively keep the previous value (assuming the default 60s was not overriden).

thallgren replied:
No, it actually changes value from 15s to 30s :-) (and makes it dependent on the `trafficManagerConnect`).

P0lip replied:
ah yeah, I meant 60 (the default value as I mentioned there) divided by 2 which is indeed 30. I don't know why but I saw `15*time.Second` as `2*15*time.Second` in my head

P0lip commented on docs/reference/client.md:
```diff
@@ -13,15 +13,24 @@ The [Telepresence CLI client](../quick-start.md) is used to connect Telepresence
+| `connect`        | Starts the local daemon and connects Telepresence to a namespace in your cluster. After connecting, outbound traffic is routed to the cluster so that you can interact with services as if your laptop was another pod (for example, curling a service by it's name)                                                                                                                       ...
+| `curl`           | curl using a containerized executable that shares the network established by a connect. Especially useful when using `connect --docker`.                                                                                                                                                                                                                                                   ...
+| `docker-run`     | run a docker image in a container that shares the network established by a connect.  Especially useful when using `connect --docker`.                                                                                                                                                                                                                                                      ...
+| `gather-logs`    | Gather logs from traffic-manager, traffic-agents, user, and root daemons, and export them into a zip file that can be shared with others or included with a github issue. Use `--get-pod-yaml` to include the yaml for the `traffic-manager` and `traffic-agent`s. Use `--anonymize` to replace the actual pod names + namespaces used for the `traffic-manager` and pods containing `traf ...
```
```suggestion
```

thallgren committed:
Retain angle brackets in code blocks, but not in markdown text.

Top-level markdown is sensitive to HTML constructs and will often error
on a text that contains angle brackets. Codeblocks are different in this
respect, and will present HTML escapes for angle brackets verbatim, and
that's not what we want.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add changelog entry mentioning deprecation of `uninstall --agents` flag.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Doc fixes.

- Remove redundant and incorrect info on how to proxy outbound traffic
- Adjust the concepts/docker page to mention the new `curl` and
  `docker-run` subcommands.
- Fix some typos.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Provide setting for the traffic-manager pod hostNetwork.

Relates to #3734.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.0-rc.1

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip committed:
Update the release workflow to build Windows ARM64 client

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

thallgren committed:
Fix DNS resolver race condition.

The DNS resolver would sometimes return a false NOERROR EMPTY when
checking the cache prior to getting a lookup return from the server.
This commit ensures that the cached entry is considered invalid unless
it's received an answer.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Store negative returns in DNS cache.

This offloads the traffic-manager significantly because unless cached,
negative returns will often result in queries being endlessly
repeated due to timeouts.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip committed:
Prepare v2.21.0-rc.2

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

thallgren committed:
Fix deadlock when closing port-forward streams.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Increase timeout in Test_LargeRequest from 30 to 60 seconds

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix misaligned YAML in reference docs.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip committed:
Do not test Windows arm64 client if runner arch is not arm64

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

thallgren committed:
Add entry about performance improvement in list command.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix telepresence connect confusion caused by /.dockerenv file

A `/.dockerenv` will be present when running in a GitHub Codespaces
environment. That doesn't mean that telepresence cannot use docker, or
that the root daemon shouldn't start.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix lint glitch.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Move misplaced import of Platform component.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update all tools dependencies and fix subsequent lint errors.

Replaces #3742

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip committed:
Free SID after use in isAdmin

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

P0lip committed:
Prepare v2.21.0-rc.3

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

P0lip committed:
Add //nolint:errcheck to windows.FreeSid call.

Signed-off-by: Jakub Rożek <jrozek@datawire.io>

thallgren committed:
Remove obsolete code checking the Docker Bridge for DNS

The DNS resolver checked the Docker bridge for messages on Linux. This
code was obsolete and caused problems when running in Codespaces.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prepare v2.21.0-rc.4

Signed-off-by: Thomas Hallgren <thomas@tada.se>
